From 26834fab18e9275c3b535f59bf1d5176524f10fe Mon Sep 17 00:00:00 2001
From: Bharathraj Nagaraju <bnagar@motorola.com>
Date: Wed, 3 Jun 2015 05:54:01 -0500
Subject: [PATCH 259/959] IKSWL-20305: power: smbchg-mmi: Correct Battery
 status

When weak charger detected there could be situation that battery
charging current could be less than step chrg iterm current.
This causes the battery status to be reported as FULL for a while
even when reported battery SOC less than 100.

Fix this by check battery SOC before reporting battery status.

Change-Id: Ib65680249c6e0c394a743a2ee3aae64bf727c801
Signed-off-by: Bharathraj Nagaraju <bnagar@motorola.com>
Reviewed-on: http://gerrit.mot.com/753902
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Kenneth Kessler <kennykessler@motorola.com>
Reviewed-by: Jason Knopsnyder <jasonk@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/power/qpnp-smbcharger-mmi.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/power/qpnp-smbcharger-mmi.c b/drivers/power/qpnp-smbcharger-mmi.c
index 3c83a74e630..7c1c082f74e 100644
--- a/drivers/power/qpnp-smbcharger-mmi.c
+++ b/drivers/power/qpnp-smbcharger-mmi.c
@@ -308,6 +308,7 @@ enum wake_reason {
 
 static void smbchg_rate_check(struct smbchg_chip *chip);
 static void smbchg_set_temp_chgpath(struct smbchg_chip *chip, int prev_temp);
+static int get_prop_batt_capacity(struct smbchg_chip *chip);
 
 static int smbchg_debug_mask;
 module_param_named(
@@ -794,6 +795,9 @@ static int get_prop_batt_status(struct smbchg_chip *chip)
 	int rc, status = POWER_SUPPLY_STATUS_DISCHARGING;
 	u8 reg = 0, chg_type;
 	bool charger_present, chg_inhibit;
+	int batt_soc;
+
+	batt_soc = get_prop_batt_capacity(chip);
 
 	rc = smbchg_read(chip, &reg, chip->chgr_base + RT_STS, 1);
 	if (rc < 0) {
@@ -805,8 +809,8 @@ static int get_prop_batt_status(struct smbchg_chip *chip)
 	    (chip->temp_state == POWER_SUPPLY_HEALTH_GOOD))
 		return POWER_SUPPLY_STATUS_FULL;
 
-	if ((chip->stepchg_state == STEP_FULL) && !chip->demo_mode &&
-	    (chip->temp_state == POWER_SUPPLY_HEALTH_GOOD))
+	if ((chip->stepchg_state == STEP_FULL) && !(batt_soc < 100) &&
+	    !chip->demo_mode && (chip->temp_state == POWER_SUPPLY_HEALTH_GOOD))
 		return POWER_SUPPLY_STATUS_FULL;
 
 	charger_present = is_usb_present(chip) | is_dc_present(chip);
-- 
2.11.0

